src/parser.cpp

```diff
@@ -65,24 +65,21 @@ namespace ada::parser {
     // Keep running the following state machine by switching on state.
     // If after a run pointer points to the EOF code point, go to the next step.
     // Otherwise, increase pointer by 1 and continue with the state machine.
-    for (; input_position <= input_size; input_position++) {
+    while (input_position <= input_size) {
```

Comment:
So by not unconditionally incrementing input_size at the end of the iteration, we avoid the messy business of having to often decrement it if we want to stay in place. This makes it easy to have a forward-only implementation... which is nicer (you don't end up with input_position == -1).